### DISCUSSION TITLE
How to list/display popular post ?

### DISCUSSION THREAD

**Discussion by lucius100** [pocketbase/pocketbase#4012] [Q&A] [answered]
Right now, I can only list my post from latest to oldest, and I am trying to implementing top/popular post by adding view count property inside the collection, but we need to increment the viewcount by accessing with :id
can you give me example code on how to do it ?
I seems out of idea since the id is randomly generated.

**Comment by ganigeorgiev** [comment, marked as answer]
This really will depend on your collections structure and desired data flow.

Currently we don't have autoincrement field if that is what you asking.

If you want to increment a `number` field programmatically when someone sends a view request for a record, you can attach for example to the `OnRecordViewRequest` hook and update the record from there. 

The caveat when incrementing counters manually is dealing with races due to concurrent requests execution and that usually will require either wrapping the select+update queries in a transaction OR maybe send a direct raw update SQL query. Here is an example with both approaches using the JS pb_hooks (_I haven't tested it, treat it as pseudo-code_):

```js
// pb_hooks/main.pb.js

// increment a "totalViews" number field every time when someone send a view request
onRecordViewRequest((e) => {
    // v1 - wrap in a transaction
    // (mixed transactions could have slight impact on the performance because they may trigger the SQLITE_BUSY timeout!)
    // ---------------------------------------------------------------
    $app.dao().runInTransaction((txDao) => {
        // refetch to make sure that we have the latest state inside the transaction
        const record = txDao.findRecordById(e.record.collection().id, e.record.id)

        // increment the number field
        record.set("totalViews", record.get("totalViews") + 1);

        // persist the change
        txDao.saveRecord(record);

        // return in the response the latest record state
        e.record = record;
    })

    // v2 - execute a raw UPDATE SQL query
    // (this will not trigger the realtime change event for the incremented counter!)
    // ---------------------------------------------------------------
    // increment the field
    $app.dao().db()
        .newQuery("UPDATE yourCollection SET totalViews = totalViews + 1 WHERE id = {:id}")
        .bind({ "id": e.record.id })
        .execute();

    // refetch to make sure that we have the latest state
    e.record = $app.dao().findRecordById(e.record.collection().id, e.record.id);
    
    // v3 - if the counter doesn't need to be precise and don't care about races, you can just update it
    // ---------------------------------------------------------------
    e.record.set("totalViews", e.record.get("totalViews") + 1);
    $app.dao().saveRecord(e.record);
}, "yourCollection")
```

Then on the client-side, to list records based on the `totalViews` field:
```js
const result = await pb.collection("yourCollection").getList(1, 10, {
    sort: "-totalViews" // + or - for ASC or DESC order
})
```

**Reply by lucius100** [reply]
I am hesitated to use this feature, I read that while a process writing into database, it automatically lock and prevent other unable to write,   if 1 url of my website accessed concurrently by 100 users, does the increment feature will work as expected in this case ?

**Reply by ganigeorgiev** [reply]
SQLite allows only a single writer at a time and writes are queued up.
Please read the comments in the above 3 versions. If you want to have the minimal impact on the performance use v2 or v3.

**Reply by guilhermeguidetti** [reply]
@ganigeorgiev  I don't use view requests; I use List/Search to retrieve a list of posts, and I want to count the views for these posts in that list. Is there a specific property or method I can use to achieve this?

**Reply by ganigeorgiev** [reply]
Similarly to the above hook, you can attach to the [`onRecordsListRequest`](https://pocketbase.io/docs/js-event-hooks/#onrecordslistrequest) hook and iterate over the `e.records` items. How and where the views counter will be stored (as a field to each record or in a separate collection) depends on your preference and specific use case.

**Reply by CovertCode** [reply]
Use cache to read most of your data.
To clear cache you can create a hook which will send http req to a endpoint which will clear cache on record update.
I am creating this for my blog. For cache I am using cloudflare and hook to clear cache on cloudflare.
